sql: turn on query cache

RaduBerinde opened the pull request:
Release note (performance improvement): A query plan cache saves a
portion of the planning time for frequent queries.

Some benchmarks with KV: https://docs.google.com/spreadsheets/d/1n818OsKTWsatbpXo5ddOJ53mL4hY-JhNJEuIlytilLc/edit#gid=0

While there is some improvement, the implicit prepare case is still nowhere close to prepare-once; I plan to benchmark and do more work on the prepare path. I think it's a good idea to enable the cache now though, so we get more time to weed out any problems.

RaduBerinde committed:
sql: turn on query cache

Release note (performance improvement): A query plan cache saves a
portion of the planning time for frequent queries.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/34454)
<!-- Reviewable:end -->

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> It does seem like there may be some kind of issue preventing us from getting close to prepare-once. I agree that we should enable for the alpha and then try to figure out what's going on.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34454#-:-LXo8LE73zauO8fGeIXq:b-intj95)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, and @rytaft)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

RaduBerinde commented:
bors r+

andy-kimball commented:
I fixed a race condition in the query cache code. The `PreparedMetadata` was being put into the cache, but then the `InferredTypes` field was being set after that by `execPrepare`.

@RaduBerinde, I put a TODO in the code for you to see if there's a better fix. For now, I made a copy of the `PreparedMetadata` struct so that it's not shared. Perhaps embedding `PreparedMetadata` in `CachedData` instead? Or changing around logic so that we don't need to set `InferredTypes` afterwards?

andy-kimball commented:
bors r-

andy-kimball committed:
sql: Fix race condition in the query cache

The PreparedMetadata was being put into the cache, but then the
InferredTypes field was being set after that by execPrepare. The fix
is to make a copy of PreparedMetadata before entering it into the cache.

Release note: None

andy-kimball commented:
bors r+

andy-kimball commented:
bors r+

RaduBerinde commented:
Thanks, LGTM! I will look at moving that InferredTypes code earlier, before we add to the cache (or otherwise remove it from PrepareMetadata).

RaduBerinde commented:
Last run was a flake in TestDockerCLI, trying again.

bors r+

RaduBerinde commented:
Ah.. looks like we have to invalidate a cached memo if SafeUpdates changes. I guess I didn't run into that because updates were not planned with opt (I assume they are now?)

RaduBerinde commented:
Alternatively we could move that check to execution - seems like that's where it belongs conceptually.

andy-kimball committed:
opt: Invalidate memo if SafeUpdates changes

Do not use cached memo if the SafeUpdates setting does not match.

Release note: None

andy-kimball commented:
I added a check for it.

andy-kimball commented:
bors r+

RaduBerinde commented:
Thank you, looks great!
